Initial version

jpkrohling opened the pull request:
Review notes:

1. There are two commits on this PR. The first is the "very basic" integration, containing only an interceptor. This would be very similar to the EJB integration.
1. The second commit moves a few things around and exposes some producers. I'm not sure they will be more useful or more of a burden: if applications are often overriding our defaults, then we are better off without the producers. As they are really simple, I'd vote for *not* having those in the first version, perhaps adding them on a future version: it's easier to add stuff than to remove.
1. Please, read/review the readme as well
1. And the example
1. It requires a non-released version of the EJB integration. If you need to build this PR locally, I'll release the EJB module and bump the version here.

jpkrohling committed:
Initial implementation

jpkrohling commented on opentracing-cdi-example/README.md:
```diff
@@ -0,0 +1,66 @@
+
+The example has the following call tree:
+
+![Trace example](call-tree.png)
```
This image is still missing :)

jpkrohling commented on opentracing-cdi-example/src/main/java/io/opentracing/contrib/cdi/example/OrderService.java:
```diff
@@ -0,0 +1,29 @@
+    InventoryService inventoryService;
+
+    @Inject
+    SpanContext spanContext;
```
Note that this has a different semantic as if this had been a method parameter. For a method parameter, this is a "child" span of the caller:

```java
public void action() {
  otherService.anotherAction(this.spanContext);
}
```

The interceptor will replace the span context received by `anotherAction` by a child span of the span sent by `action`.

When injecting the `SpanContext`, we get it from the request attributes, so, it would be misleading to create this span as a direct child of the request. I think we should better let the application decide, but I'm open for suggestions.

jpkrohling commented on opentracing-cdi-example/src/main/java/io/opentracing/contrib/cdi/example/SpanContextProducer.java:
```diff
@@ -0,0 +1,18 @@
+
+    @Produces
+    public SpanContext produceSpanContext() {
+        return (SpanContext) request.getAttribute(TracingFilter.SERVER_SPAN_CONTEXT);
```
Note that this is on the example code, this is not part of the CDI integration. I'm not sure we can (or should) make the assumption that the span context is what we have on the request. Should we?

pavolloffay replied:
In the example I guess it's fine. However I guess CDI could work also without servlets

jpkrohling commented on opentracing-cdi-example/src/main/java/io/opentracing/contrib/cdi/example/SpanContextProducer.java:
```diff
@@ -0,0 +1,18 @@
+
+    @Produces
+    public SpanContext produceSpanContext() {
+        return (SpanContext) request.getAttribute(TracingFilter.SERVER_SPAN_CONTEXT);
```
Another point to discuss: should the integration provide a producer for `ActiveSpan`, for no-async applications? If so, what should it return?

```java
    @Produces
    public ActiveSpan produceActiveSpan(InjectionPoint ip) {
        return GlobalTracer.get().activeSpan();
    }
```
(this would require whoever started the span to deactivate it)

Or:
```java
    @Produces
    public ActiveSpan produceActiveSpan(InjectionPoint ip) {
        ActiveSpan parent = GlobalTracer.get().activeSpan();
        return GlobalTracer
                .get()
                .buildSpan(ip.getMember().getName())
                .asChildOf(parent.context())
                .startActive();
    }
```
(this last would *require* the place where it gets injected to `deactivate` the span)

sjoerdtalsma replied:
> (this would require whoever started the span to deactivate it)

..whoever ~~started~~ activated the span to deactivate it..
I think it is most reasonable that activation / deactivation are paired. i.e. there _could_ be an active span (or may not be). However, the one placing it there is also responsible for removing it.
(usually in a try .. with resources construct from some executorservice that propagates activespans).

sjoerdtalsma replied:
Availability of an _active_ span basically means:
Here's some current, ongoing span. If you want to create a span by start / finish, it may be beneficial to become _child of_ this ongoing span.

I would definitely _not_ get involved with the lifecycle of this parent!

jpkrohling replied:
So, the first option ?

jpkrohling commented on opentracing-cdi-example/src/main/java/io/opentracing/contrib/cdi/example/TracerInitializer.java:
```diff
@@ -0,0 +1,36 @@
+ *
+ * @author Juraci Paixão Kröhling
+ */
+public class TracerInitializer {
```
This might be generic enough and could be placed on the integration itself.

jpkrohling commented on pom.xml:
```diff
@@ -72,6 +73,16 @@
+      <dependency>
+        <groupId>io.opentracing.contrib</groupId>
+        <artifactId>opentracing-ejb</artifactId>
+        <version>0.0.3-SNAPSHOT</version>
```
This is because we have a new method on the interceptor. A new PR will be sent once this PR here gets into an advanced stage.

pavolloffay replied:
I wanted to compile and play with this.

sjoerdtalsma replied:
Yeah, not merge-able while SNAPSHOT dependencies exist.
Will travis build even?

jpkrohling replied:
Released and updated.

jpkrohling commented:
I'd also appreciate feedback from maintainers of other Java-related integration repositories, like @sjoerdtalsma and @malafeev

pavolloffay commented on opentracing-cdi/pom.xml:
```diff
@@ -22,8 +22,21 @@
+    </dependency>
+    <dependency>
+      <groupId>io.opentracing.contrib</groupId>
+      <artifactId>opentracing-web-servlet-filter</artifactId>
```
Why it depends on this? CDI might be used without servlets.

jpkrohling replied:
True, but if we want the servlet integration, we need to create a producer that gets data from the request context, and this requires this dependency. Should I remove the SpanContext producer?

pavolloffay replied:
What about using ActiveSpanSource?

pavolloffay commented on opentracing-cdi/pom.xml:
```diff
@@ -22,8 +22,21 @@
     </dependency>
+    <dependency>
+      <groupId>io.opentracing.contrib</groupId>
+      <artifactId>opentracing-ejb</artifactId>
```
Why it depends on EJB?

jpkrohling replied:
The base interceptor is there. I'm thinking about extracting the interceptor into a `java-interceptor`, and making CDI and EJB use it, but I'm not convinced yet. The EJB dependency brings nothing with it, except the name "EJB", so, as of now, I'll keep it like this.

pavolloffay replied:
> java-interceptor 

or just a separate package in java-ejb or here. I don't know what is more appropriate.

pavolloffay commented on opentracing-cdi-example/README.md:
```diff
@@ -0,0 +1,69 @@
+
+## Running
+
+In order to visualize the spans, you'll need an instance of Jaeger running locally.
```
It would be maybe better to use `MockTracer` and send spans to stdout. At least we have used this in other examples in this org.

jpkrohling replied:
This also sends spans to stdout *and* provides a UI :)

objectiser replied:
Agree with Pavol - the ot-contrib repos need to be vendor neutral :)

jpkrohling replied:
Even for the examples? Would it be OK if I changed this to mention that the examples can also be run with *any* other OpenTracing-compatible tracers?

objectiser replied:
My preference would be for being tracer neutral, but you could ask wider OT community to get other opinions.

pavolloffay reviewed (commented):
Can this work without depending on servlets and EJB?

sjoerdtalsma commented on README.md:
```diff
@@ -4,6 +4,113 @@
+
+### Tracer registration via  GlobalTracer
+
+This integration also features a registration of the `Tracer` with the `GlobalTracer`,
```
Maybe add a remark about what happens when there are conflicting `@Producer` statements and an explicitly-registered `GlobalTracer`?
The `GlobalTracer` currently has register-once semantics so cannot be re-registered.

jpkrohling replied:
Done

pavolloffay commented on opentracing-cdi/src/main/java/io/opentracing/contrib/cdi/internal/TracerInitializer.java:
```diff
@@ -0,0 +1,60 @@
+
+            if (!(tracer instanceof GlobalTracer)) {
+                // it's not our own producer, which returns a GlobalTracer. We'll just use it then!
+                GlobalTracer.register(tracer);
```
Integration should not register the tracer.

jpkrohling replied:
This is explained in the readme file. Do you disagree with what is explained there? What would be the right procedure?

pavolloffay replied:
About the registration. If multiple instrumentations try to register the tracer it will throw a runtime exception.

sjoerdtalsma replied:
Minor nuance; if multiple instrumentations try to register _the same_ tracer it will be lenient.

objectiser replied:
May be better to move this to a separate artifact under the java-cdi repo, related to tracer resolution. So then there would be artifact for instrumentation of CDI components, and a separate artifact to provide access to a tracer via CDI.

jpkrohling replied:
There's a flag on this logic to skip this procedure. But if an artifact is more useful, I can extract it then (just not sure it deserves that much).

objectiser replied:
Just thinking it would be cleaner to distinguish instrumentation of CDI, versus obtaining a Tracer using CDI. Although the instrumentation of classes/methods must be explicitly marked, so I guess having them in separate artifacts is not necessary.

sjoerdtalsma commented on opentracing-cdi-example/src/main/java/io/opentracing/contrib/cdi/example/AccountService.java:
```diff
@@ -0,0 +1,20 @@
+ * This is a simple synchronous EJB, without any knowledge about span context or other OpenTracing semantics. All it
+ * does is specify an interceptor and it's shown as the child of a parent span.
+ *
+ * @author Juraci Paixão Kröhling
```
Maybe encode the non-ascii unicode characters?
I have been bitten many times with source files edited by different people using different IDE's on different operating systems.

Opinions on `@author` vary; @adriancole has asked me in previous reviews to drop the tag altogether.

jpkrohling replied:
Doesn't Java defaults to UTF-8? I personally never had problems with that in Java.

About the `@author` tag: I don't mind dropping it, and I certainly don't do it for "credits", but I did find it useful in the past to contact the original author of a given code, to ask for some clarification.

pavolloffay commented on opentracing-cdi/src/main/java/io/opentracing/contrib/cdi/internal/TracerInitializer.java:
```diff
@@ -0,0 +1,60 @@
+ *
+ * @author Juraci Paixão Kröhling
+ */
+public class TracerInitializer {
```
I think the order of getting the tracer should be:

1. get tracer from CDI context
2. get from globalTracer
(3. use tracerresolver) - if we decide to depend on it

jpkrohling replied:
That's exactly what it's doing ;-)

pavolloffay replied:
I think there might be a problem when multiple instrumentations are being used together and all of them are using tracerresolver. 

It would be maybe better if the tracer artifact which supports tracerresolver would resolve tracer and register it to the GlobalTracer. Instrumentations would just depend on GlobalTracer.

sjoerdtalsma replied:
:+1:  I think that would simplify things.

jpkrohling replied:
So, should I register only if there's an producer for `Tracer` that returns something else than `GlobalTracer`?

pavolloffay replied:
I think you should not register at all

sjoerdtalsma commented on opentracing-cdi-example/src/main/java/io/opentracing/contrib/cdi/example/InventoryNotificationService.java:
```diff
@@ -0,0 +1,33 @@
+    Tracer tracer;
+
+    @Inject
+    ActiveSpan parentSpan;
```
Why `parentSpan` and not `activeSpan` ?
It may be that you're using it for the `child-of` relation below, but maybe add a comment on this injection?

sjoerdtalsma replied:
Also, this service starts throwing exceptions when there is either no `tracer` or `parentSpan`.
Of the _tracer_ I can understand that an application developer chooses to always have one (with easy fallback to the noop tracer obviously).
For the _parent span_ I'm not so sure. You now have a service class that starts throwing exceptions on untraced call paths. For the example it may be okay, but maybe explain how you could make this work without an active span?

jpkrohling replied:
See https://github.com/opentracing-contrib/java-cdi/pull/1#discussion_r130872419

pavolloffay commented on README.md:
````diff
@@ -4,6 +4,113 @@
+```
+
+If you have your own `Tracer` producer, you'll need to mark your own as an
+`@Alternative`
````
I think alternative means something like not primary. Could we annotate our producer(in instrumentation) with `@Alternative`.

I want to say that user provided tracer should be the primary one and not an alternative.

jpkrohling replied:
`@Alternative` for CDI means "this is more important" (if listed on `beans.xml`, that is). See http://docs.oracle.com/javaee/6/tutorial/doc/gjsdf.html

pavolloffay replied:
It's weird because they are using it for test impl. Other thing is that alternative bean has to be defined in `beans.xml` Am I right? If this is true it makes it harder to configure.

```
<beans ... >
    <alternatives>
        <class>encoder.TestCoderImpl</class>
    </alternatives>
</beans>
```

Is there a way to make user defined producer as a primary? The best without any additional configuration.

jpkrohling replied:
The options, as I see it, are:

- Have a non-qualified producer for `Tracer`, requiring developers to either qualify their own tracer or specify an alternative via `beans.xml`
- Have a qualified producer for `Tracer`, requiring developers to use this qualifier every time they want to use our producer.

I believe that our producer will be the most commonly used, so, the easiest for most developers will be to just have `@Inject Tracer tracer` and get a Tracer as returned by `GlobalTracer.get()`.

I'm preparing some tests, so that it's clear what are the options.

pavolloffay replied:
> Have a qualified producer for Tracer, requiring developers to use this qualifier every time they want to use our producer.

Do you mean they would have to use that annotation when injecting the tracer? 

> I believe that our producer will be the most commonly used,

I would personally write my own producer which would return a configured tracer

jpkrohling replied:
> Do you mean they would have to use that annotation when injecting the tracer?

Yes.

> I would personally write my own producer which would return a configured tracer

Will that be the common case?

pavolloffay replied:
> Will that be the common case?

Hard to say, but maybe if people are in CDI environment they might prefer it.

About alternative etc. I think we should just use the simplest and most intuitive solution.

jpkrohling replied:
> About alternative etc. I think we should just use the simplest and most intuitive solution.

This all depends on what's the common case. The simplest and "most intuitive" solution is to just let people add stuff as maven dependency and consume a tracer via `@Inject Tracer tracer`. All boiler plate code, like creating a new tracer, would be done by "something" (TracerResolver, CDI integration, ...). It becomes a problem if the common case is for people to specify their own `Tracer`.

For my demos and examples, I never create a Tracer by myself: I always use system properties to configure Jaeger's Tracer, and use Jaeger's TracerResolver module, so that the tracer of choice becomes more like a packaging/runtime decision and less of a developer's decision.

sjoerdtalsma commented on opentracing-cdi-example/src/main/java/io/opentracing/contrib/cdi/example/OrderPlacementEvent.java:
```diff
@@ -0,0 +1,15 @@
+public class OrderPlacementEvent {
+    private SpanContext spanContext;
+
+    public OrderPlacementEvent(SpanContext spanContext) {
```
Not sure I like the mixing of tracing concern with the business logic here.
Intuitively, the tracing context seems out of place in a business event signature to me.

jpkrohling replied:
This is actually following Google's pattern of explicit context propagation (also used on Dapper, IIRC), but users aren't required to do so. See the other event, `InventoryChangeEvent`, for a no-context example.

sjoerdtalsma commented on opentracing-cdi/src/main/java/io/opentracing/contrib/cdi/Traced.java:
```diff
@@ -0,0 +1,18 @@
+@InterceptorBinding
+@Retention(RUNTIME)
+@Target({METHOD, TYPE})
+public @interface Traced {
```
Missing `JavaDoc` on the annotation.
This will become an important (visible) part of the API.

jpkrohling replied:
Added javadocs, not only to this annotation.

sjoerdtalsma commented on opentracing-cdi/src/main/java/io/opentracing/contrib/cdi/internal/TracerInitializer.java:
```diff
@@ -0,0 +1,60 @@
+
+            if (!(tracer instanceof GlobalTracer)) {
+                // it's not our own producer, which returns a GlobalTracer. We'll just use it then!
+                GlobalTracer.register(tracer);
```
Please be aware this may throw a runtime exception if there's already a `GlobalTracer` registered (by someone else).

jpkrohling replied:
Good catch. I just moved the check about an already registered tracer to be one of the first things on this initializer.

sjoerdtalsma reviewed (commented):
Just some comments after a first quick look.

pavolloffay commented on pom.xml:
```diff
@@ -1,171 +1,231 @@
+                <groupId>junit</groupId>
+                <artifactId>junit</artifactId>
+                <version>4.12</version>
+                <scope>test</scope>
```
this is not necessary here

pavolloffay replied:
The same applies for all following dependencies

pavolloffay commented on opentracing-cdi/src/main/java/io/opentracing/contrib/cdi/internal/TracerInitializer.java:
```diff
@@ -0,0 +1,86 @@
+
+/**
+ * Registers a {@link Tracer} with the {@link GlobalTracer}, should it be a tracer that is defined by the target
+ * application. If there are no application-specific {@link Tracer} available, this initializer attempts to locate one
```
>  aplication-specific

Could you clarify? Is it tracer from globalTraer or CDI? Simple comment about the priority would be good.

pavolloffay commented on opentracing-cdi/src/main/java/io/opentracing/contrib/cdi/internal/SpanContextProducer.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Produces
+    public SpanContext produceSpanContext() {
+        SpanContext spanContext = (SpanContext) request.getAttribute(TracingFilter.SERVER_SPAN_CONTEXT);
```
CDI should not depend on servletfilter

jpkrohling replied:
The alternative is to remove this producer. I don't think CDI in Java SE is that popular, so, I'd be opinionated here and assume that people wanting to use the CDI integration for OpenTracing are going to be running on a Servlet environment.

That said, I'm OK in removing this producer, if more people think this is preferable.

pavolloffay replied:
I don't have a strong opinion about this but I think ActiveSpan producer might be enough. These to represent the same span. 

minor not in async request `request.getAttribute(TracingFilter.SERVER_SPAN_CONTEXT);` would return null and I guess active span producer too.

objectiser replied:
Agree, active span should be enough, so may not be an issue.

jpkrohling replied:
OK, I'll remove this producer then.

pavolloffay commented on opentracing-cdi/src/main/java/io/opentracing/contrib/cdi/internal/SpanContextProducer.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * Produces a {@link SpanContext}, based on the {@link HttpServletRequest}'s attributes.
+ */
+public class SpanContextProducer {
```
Why is this needed? Where do we inject `SpanContext`?

jpkrohling replied:
See:
https://github.com/jpkrohling/java-cdi/blob/5ca652c63a9e29c01f4118003265ca4705119b03/README.md#injecting-a-spancontext

and

https://github.com/opentracing-contrib/java-cdi/pull/1#discussion_r130870751

pavolloffay replied:
oh I thought that it is also used within this instrumentation

sjoerdtalsma commented:
Regarding your original comment
>2. The second commit moves a few things around and exposes some producers. I'm not sure they will be more useful or more of a burden: if applications are often overriding our defaults, then we are better off without the producers. As they are really simple, I'd vote for not having those in the first version, perhaps adding them on a future version: it's easier to add stuff than to remove.

Maybe it's an idea to remove them from this PR for now and create a separate PR for those?

jpkrohling commented:
> Maybe it's an idea to remove them from this PR for now and create a separate PR for those?

Which ones should I remove? We have producers for `SpanContext`, `ActiveSpan` and `Tracer`. I wouldn't remove the initializer right now, as it can be skipped by applications that don't want to use it.

sjoerdtalsma commented:
>>Maybe it's an idea to remove them from this PR for now and create a separate PR for those?

>Which ones should I remove? We have producers for SpanContext, ActiveSpan and Tracer. I wouldn't remove the initializer right now, as it can be skipped by applications that don't want to use it.

Yes maybe indeed keep the initializer in this PR, however, I'd like to see some 'CDI library' recognition in the property name instead of the general `"skipTracerInitializer"`. Best to avoid clashes with other instrumentation that may wish to have a system property to skip tracer initialization.

jpkrohling commented:
> Maybe it's an idea to remove them from this PR for now and create a separate PR for those?

How about we keep it for now and open an issue on this repo, to re-evaluate that pre-1.0.0? This way, we can do a survey with actual users, to measure how useful it's been.

jpkrohling commented:
I've just squashed and updated this PR.

Points that are still open:
- The discussion about the TracerResolver
- Removal or not of the "extra" producers (`SpanContext`, `ActiveSpan` and `Tracer`)
- Dependency on the `java-ejb` (which is only a "naming" problem for now)

If there aren't blocking issues with this PR, I'd like to merge it and discuss the other points in separate issues.

objectiser commented on README.md:
```diff
@@ -4,6 +4,135 @@
+All the caveats of using the `ActiveSpan` apply. As a rule of thumb, it can be considered
+safe to use on synchronous applications, but should not be used for async processing. You can
+safely mix sync and async if you pass the context from the sync object to the async explicitly,
+like:
```
well there is the `Continuation` mechanism - but as this may be changing in the near future, maybe this can be left for now.

objectiser commented on opentracing-cdi/src/main/java/io/opentracing/contrib/cdi/internal/SpanContextProducer.java:
```diff
@@ -0,0 +1,47 @@
+        ActiveSpan activeSpan = tracer.activeSpan();
+        if (null == activeSpan) {
+            String spanName = ip.getMember().getName();
+            return tracer.buildSpan(spanName).startActive();
```
Need to also add the component tag (and any other, e.g. 'class').

jpkrohling commented:
As I understand that there are no blockers, I'm merging this and opening issues on this repository to discuss the individual points.
